### ISSUE TITLE
missing flock on unsupported OS (solaris)

### ISSUE THREAD

**Issue by siepkes** [sorintlab/stolon#232]
I tried to compile stolon for use on SmartOS (Illumos / Solaris) an ran into an issue. The issue is that CoreOS rkt (on which Stolon is dependent) uses the `flock` system call. Apparently Illumos now supports flock (see: [Feature #3252](https://www.illumos.org/issues/3252)) however to Go Solaris library does not yet support it. Long story short; Solving this upstream in Go is obviously the right way (which is apparently also being worked on) but thats going to take a looooooong time before being complete. 

Now I noticed that the only reason why stolon is (currently) dependent on rkt is because of a lock implementation which is used in `keeper.go` line 1406 and nothing else. Would you be open to ditching the rkt dependency and using a lock implementation which also works on Illumos / Solaris? rkt also seems like quite a big "library" to use for just a small lock implementation.

For Google-ing reference, this is the error Go dies with when trying to compile Stolon (with rkt) for Solaris:
```
# github.com/sorintlab/stolon/vendor/github.com/coreos/rkt/pkg/lock
gopath/src/github.com/sorintlab/stolon/vendor/github.com/coreos/rkt/pkg/lock/file.go:49: undefined: syscall.Flock
gopath/src/github.com/sorintlab/stolon/vendor/github.com/coreos/rkt/pkg/lock/file.go:49: undefined: syscall.LOCK_EX
gopath/src/github.com/sorintlab/stolon/vendor/github.com/coreos/rkt/pkg/lock/file.go:49: undefined: syscall.LOCK_NB
gopath/src/github.com/sorintlab/stolon/vendor/github.com/coreos/rkt/pkg/lock/file.go:75: undefined: syscall.Flock
gopath/src/github.com/sorintlab/stolon/vendor/github.com/coreos/rkt/pkg/lock/file.go:75: undefined: syscall.LOCK_EX
gopath/src/github.com/sorintlab/stolon/vendor/github.com/coreos/rkt/pkg/lock/file.go:96: undefined: syscall.Flock
gopath/src/github.com/sorintlab/stolon/vendor/github.com/coreos/rkt/pkg/lock/file.go:96: undefined: syscall.LOCK_SH
gopath/src/github.com/sorintlab/stolon/vendor/github.com/coreos/rkt/pkg/lock/file.go:96: undefined: syscall.LOCK_NB
gopath/src/github.com/sorintlab/stolon/vendor/github.com/coreos/rkt/pkg/lock/file.go:122: undefined: syscall.Flock
gopath/src/github.com/sorintlab/stolon/vendor/github.com/coreos/rkt/pkg/lock/file.go:122: undefined: syscall.LOCK_SH
gopath/src/github.com/sorintlab/stolon/vendor/github.com/coreos/rkt/pkg/lock/file.go:122: too many errors
```

**Comment by sgotti** [comment]
@siepkes the usage of a package provided by rkt is not an issue and doesn't mean that this project depends on rkt, if you look at the vendor dir we are just vendoring the rkt/pkg/lock (3 files) because I know it's reliable since I contributed to its development. It could be easily extracted from the rkt repo (there's a related issue in the rkt project).

The real issue is just that it's made to work on GNU/linux systems (it will require a build tag to better report thay it's not supported on other OSes) using the flock syscall because fcntl is not very reliable in some cases. So the fix will be to support other oses different than linux.

For the locking it's only needed by the keeper to lock the data dir so another library that supports multiple OSes like https://godoc.org/go4.org/lock (but should prefer flock on linux) could be used.


About a general support for opensolaris there could probably be other issues, so to really support it we need that all the continuous integration tests can be executed also in these OSes and someone willing to maintain the support for it (since we don't have the time for doing this).

**Comment by siepkes** [comment]
@sgotti  So you mean replacing the lock with a "simple" call to `fcntl` via the Go syscall package?

Besides that it is definitely good practice to run the integration tests on supported platforms do you anticipate any problems when running stolon on a platform like SmartOS (Solaris)? It's something I'm interested in but I cannot really over see what I'm getting in to here ;-)

**Comment by sgotti** [comment]
@siepkes  Sorry I wasn't clear, I meant, since the lock is just used by the keeper to lock the datadir at start, doing this lock with fcntl should be ok since it's not affected by fcntl downsides (closing any other file descriptor for the same file will release the lock prematurely). So I was suggesting to use something like go4.org/lock if it work with solaris 

>  do you anticipate any problems when running stolon on a platform like SmartOS (Solaris)? It's something I'm interested in but I cannot really over see what I'm getting in to here ;-)

We never tested this on something other than GNU/Linux. Probably the unique OS related code I can think of is the external process execution done by the keeper to run postgres (but it uses os/exec so it should wok).

**Comment by siepkes** [comment]
@sgotti I thought it could perhaps be as simple as the snippet below which I conjured up. But you seem to have a good grasp on locking on Posix / *nix so I welcome all feedback.

```
f, err := os.Open(cfg.dataDir)
if err != nil {
	fmt.Printf("Cannot open data dir %q: %v\n", cfg.dataDir, err)
	os.Exit(1)
}
	
// Configure lock to cover entire file.
ft := &syscall.Flock_t{
	Whence: io.SeekStart,
	Start:  0,
	Pid:    int32(os.Getpid()),
}

// Take an exclusive lock on dataDir	
err = syscall.FcntlFlock(f.Fd(), syscall.F_WRLCK, ft)
if err != nil {
	fmt.Printf("cannot take exclusive lock on data dir %q: %v\n", cfg.dataDir, err)
	os.Exit(1)
}

log.Info("exclusive lock on data dir taken")
```

This compiles for both Linux and Solaris. Or would you rather use the `lockPortable` in `go4.org/lock` ? As you stated we should be fine from `fcntl`'s shortcommings and `fcntl` is Posix so it's pretty portable I think. A major disadvantage of `lockPortable` is that it doesn't deal gracefully with unclean shutdowns (leaves lock files lingering).

**Comment by siepkes** [comment]
@sgotti Ping :-) If you have time, could you take a look at my previous comment? Thanks in advance!

**Comment by sgotti** [comment]
@siepkes Sorry I was on vacation. I was thinking about using `go4.org/lock` since it was much used and tested than a custom solution. The `lockPortable` will only be used on non posix systems (that stolon is not supporting and probably rather not building) and probably there is a way of disabling it.

Another solution, given the smaller usage of the locking inside stolon, will be I will to create a custom implementation like yours. Instead of polluting the keeper code It should live inside `pkg/lock` and have the current filename extensions/build tags to work on posix systems and return an error on other architectures. It should also provide some tests (also if the keeper locking is already tested here: https://github.com/sorintlab/stolon/blob/master/tests/integration/init_test.go#L454 ). But this will be more work for the unique gain of removing the portableLock...

I'm open for both solution and will be happy to review a PR.

**Comment by sgotti** [comment]
Fixed in #236
